requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+
+This RFC proposes both a technical and product solution to this problem: Store `person_id` and `person_properties` on the main events table.
+
+However, doing so will require breaking changes to the product. These changes are likely to cause some issues for customers, but these have either a negligible impact (affecting around 0.03% of events) or can be mitigated through documentation and education.
+
+## Proposed solution
+
+From a technical standpoint: Store `person_id` and `person_properties` on the main events table.
+
+From a product standpoint storing person_id and person_properties on events requires making the following changes:
+1. After the first session we will no longer merge events prior to login with the main person’s events.
+2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+### Customer Impact
+
+#### Benefits
+For both changes the benefits are as follows:
+Significant short term performance gains
+Ability to support the largest tiers of customer now and in the future
+Standard analysis such as events from paying users over time will stop yielding misleading results when people start / stop paying (as they do today)
+
+
+#### 1. After the first session we will no longer merge events prior to login with the main person’s events.
+
+![image](https://user-images.githubusercontent.com/85295485/151368075-a7ab9dc7-ee8d-4cad-a454-2524161973f1.png)
+
+How we'll map persons to events going forward:
+1. On initial visit, users device gets assigned an an (anonymous) distinct id `anonid1`. Posthog internally associates person_id `p1` to `anonid1`, which is saved together with their events
+2. When after signing up `posthog.identify('userid')` is called, we associate the same person_id `p1` with the new distinct id `userid`
+3. If the user later visits the site with a new device, a new anonymous distinct id `anonid2` is created. Posthog internally associates these events with a new person id `p2` as no mapping for that id exists.
+4. If after signing in `posthos.identify('userid')` is called, we associate events going forward with person_id `p1`
+
+As a result, events done between steps 1, 2 and 4 will be tracked as from person `p1` but events between steps 3 and 4 will be under person`p2`. As a result no events are lost but some will be considered as coming from a different (anonymous) person. In the above image this corresponds to the green period.
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
+
+As a mitigation, if necessary, we would consider a buffer mechanism to reduce affected events to ~0.003%, we can provide further details on this solution, however we would ideally avoid implementing this as it increases the complexity of the system quite significantly.
```

Comment:
(reposting a new thread from an earlier comment)

Could we have some way to "backtrack" to the anonymous events/session post login?

For example, by adding the property `anon_distinct_id: "anonid2"` to events `11` and `12` in the table above. 

This would allow us to link session recordings with the pre-login part, and potentially allow some flexibility in the future when someone asks for a funnel one part of it is logged out events, and the other part consists of logged in events.

Reply:
TLDR: what if we delayed ingestion storing the anonymous user events somewhere and ingesting after identify event with the logged_in person ID. Then we'd still get our perf and simplicity improvements for queries without losing pre-logged in events being tied to the same person.

Long version:
Assumptions: Pre-login we don't know the logged_in person_id, hence events will be sent with a new uuid

There are only 3 options (let me know if you can think of any more):
### 1. join persons 
This is what we do now & has all the perf problems for queries. If we are not querying over a single person_id for unique users queries results in us joining persons. Adding any properties etc results in the same need to join persons. I'd bucket any ideas with running queries across CH/postgres here too & I think they all have the same perf problem.

#### 2. update events in clickhouse 
Updating the person_id from the anonymous one to the logged_in person_id: We can't really mutate events in CH, what we could do is run an async migration on the shard (essentially copy all data to a new table modifying the distinct_id's accordingly & then rename tables like 0002 async migration). This is probably expensive in terms of resources even done for a single shard; that migration would put us temporarily in a state of delayed ingestion (while we copy the last part of the table and rename we'd be queueing up new events in kafka similar to 0002).
#### 3. delay ingestion until we know the logged_in person_id: 

There are two options here which might not be all that bad of an idea:
1. delay by using some intermediary storage for logged_out users: we could dump all anonymous events somewhere keyed by uuid and then consume them once we see the identify event (I was thinking kafka, but we can't get messages by key, maybe by some separate table in ClickHouse) and we we can write the events to ClickHouse with the logged_in person_id. I was thinking up to certain delay, e.g. 1 h or 1 day could be longer in theory too, but note that during that delay we wouldn't see those events in queries (which could be acceptable). I wonder if this would actually not be a terrible idea?
2. Re-ingest the events with the new person_id and mark the other uuid as a deleted person (do we filter out deleted people somehow in queries or in any case filtering out is potentially cheaper than joins). We can occasionally run an async migration that clears them out.

-----
 
Additionally there's a problem tied to person properties and how they would get updated, but we could just decide on a global semantics that person properties can only be used on logged_in users (in any case this is probably less important and the funnels working etc is more important).

Reply:
This is pretty interesting. In a way, this is what the `session_id` is today.